mypy/checkmember.py

```diff
@@ -314,7 +316,9 @@ def analyze_var(name: str, var: Var, itype: Instance, info: TypeInfo, node: Cont
                 # methods: the former to the instance, the latter to the
                 # class.
                 functype = t
-                check_method_type(functype, itype, var.is_classmethod, node, msg)
+                # Use meet to simulate dispatch - e.g. reduce Union[A, B] to A on dispatch to A
```

Comment:
The comment is unclear. Can you add a comment with an example where this makes a difference?

Reply:
I added a comment and a test. This kind of dispatch with not-exactly-the-right-type occur in reverse operators.

I believe the same thing should also be done with plain functions, but it's not checked at all now. i.e. there should be a call to `check_self_arg` before `bind_self` [here](https://github.com/python/mypy/blob/master/mypy/checkmember.py#L93). For example:
```python
class A:
    def f(self: B) -> None: pass  # type: ignore - because reasons. probably useful for metaclasses

class B(A): pass
A().f()  # should be an error
```
but changing _plain functions_ breaks some typeshed code (specifically, AbstractSet operators and the `version_info` tests break).